verifying/verifying.go

```diff
@@ -75,34 +79,68 @@ func Verify(p *shared.Proof, m *shared.ProofMetadata) error {
 		return fmt.Errorf("invalid indices set size; expected %d, given: %d", expectedSize, len(p.Indices))
 	}
 
-	difficulty := shared.ProvingDifficulty(numLabels, uint64(m.K1))
+	if options.verifyFunc == nil {
+		difficulty := shared.ProvingDifficulty2(numLabels, m.B, m.K1)
+		options.logger.Debug("verifying difficulty %d", difficulty)
+		options.verifyFunc = func(val uint64) bool {
+			return val < difficulty
+		}
+	}
+
 	buf := bytes.NewBuffer(p.Indices)
 	gsReader := shared.NewGranSpecificReader(buf, bitsPerIndex)
-	indicesSet := make(map[uint64]bool, m.K2)
+	indicesSet := make(map[uint64]struct{}, m.K2)
+
+	// create the ciphers for the specific nonce
+	d := shared.CalcD(numLabels, m.B)
+	offset := p.Nonce * uint32(d)
+	nonceBlock := uint8(offset / aes.BlockSize)
+
+	// since the value can be on a boundary between two blocks, we need to create two ciphers
+	ciphers := make([]cipher.Block, 2)
+	for i := uint8(0); i < 2; i++ {
+		c, err := oracle.CreateBlockCipher(m.Challenge, nonceBlock+i)
+		if err != nil {
+			return fmt.Errorf("creating cipher for block %d: %w", nonceBlock, err)
+		}
+		ciphers[i] = c
+	}
+
+	block := make([]byte, aes.BlockSize)
+	out := make([]byte, aes.BlockSize*2)
+	u64 := unsafe.Slice((*uint64)(unsafe.Pointer(&out[offset%aes.BlockSize])), 1)
```

Comment:
How about creating a pointer to uint64 instead of []uint64 and using it as `*u64` instead of `u64[0]`.
```suggestion
	u64 := (*uint64)(unsafe.Pointer(&out[offset%aes.BlockSize]))
```